### DISCUSSION TITLE
Are schemas loaded via $ref added to any working namespaces?

### DISCUSSION THREAD

**Discussion by JC3** [json-schema-org/community#197] [Q&A]
Consider the following two schema documents:

Document 1: Retrieval URI irrelevant:

```
{
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": "xri://+this.discussion/schema1",
    "title": "Example Schema 1",
    "properties": {
        "someProperty": {
            "$ref": "file:///path/to/local-schema.json#/$defs/def1"
        }
    }    
}
```

Document 2: Local filesystem at */path/to/local-schema.json*:

```
{
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": "xri://+this.discussion/schema2",
    "title": "Example Schema 2",
    "$defs": {
        "def1": {
            "somekeyword": "somevalue",
            "def1b": { "somekeyword": "somevalue" }
        },
        "def2": { "somekeyword": "somevalue" }
    }
}
```

Now, make the following assumptions about an application that is processing schemas:

1. Document 1 is the document of interest, and also the only specific document that the application ["understands ahead of time"](https://json-schema.org/draft/2020-12/json-schema-core.html#name-loading-a-referenced-schema).
2. The application implements a policy of treating missing reference URIs as URLs and attempting to retrieve them; note that this is a JSON Schema spec-compliant behavior. 

So, when processing document 1, the application sees the reference to `file:///path/to/local-schema.json#/$defs/def1`, notes that it is not an already-known identifier and so treats it as a URL, retrieves document 2, resolves the fragment, all is well.

However, my question is: upon retrieving document 2, should document 2 now exist in the space of schemas that the application is aware of (...even if only in some sort of limited scope)? 

For example, if document 1 were, instead, the following:

```
{
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": "xri://+this.discussion/schema1",
    "title": "Example Schema 1",
    "properties": {
        "someProperty": {
            "$ref": "file:///path/to/local-schema.json#/$defs/def1",
            "newSubkey1": { "$ref": "schema2#/$defs/def1" },
            "newSubkey1b": { "$ref": "schema2#/$defs/def1/def1b" },
            "newSubkey2": { "$ref": "schema2#/$defs/def2" }
        },
        "newProperty1": { "$ref": "schema2#/$defs/def1" },
        "newProperty1b": { "$ref": "schema2#/$defs/def1/def1b" },
        "newProperty2": { "$ref": "schema2#/$defs/def2" }
    }    
}
```

Should any of those other reference work? They would all resolve to the absolute URI `xri://+this.discussion/schema2` with fragments, which matches document 2. 

I have follow up questions that depend on whatever the answer is here, I'll leave them out for now.

**Comment by karenetheridge** [comment]
> Should any of those other reference work? They would all resolve to the absolute URI xri://+this.discussion/schema2 with fragments, which matches document 2.

It depends on which of the `$ref` or `properties` keywords are evaluated first. schema2 (and the $id it defines) aren't known to the evaluator until the document referenced by `$ref` is loaded and parsed.

Therefore, the evaluation result depends on keyword execution order, which is not well-defined by the specification.

The only exceptions to the indeterminate ordering are not well-stated in the spec, but are implicit from property definitions:
- $schema must be evaluated first, because its result will affect all the evaluation semantics (that is, the dialect is determined here);
- $id must be evaluated next, as it determines the base URI for future $ref and $dynamicRef evaluations, as well as the base URI used for in `absoluteKeywordLocation` in all errors and annotations;
- the unevaluated\* keywords must be evaluated after the applicator keywords that produce annotations that they consume.

IMO the spec should clearly state that all of the core vocabulary keywords MUST be evaluated first before all other vocabularies, and then the $ref ambiguity described by this thread will be resolved.

**Reply by JC3** [reply]
Thanks. Confirming a point: When you say it depends on the evaluation order, that implies that the answer at least isn't "no" -- i.e. evaluation order aside, schema documents loaded by $ref *should* (at some unspecified point) get added to the set of known schemas and stay there for the remainder of the evaluation process. Is there anything in the current draft that implies that this should happen at *all*, or is it completely unspecified?

**Reply by JC3** [reply]
(Btw, re: your last sentence; should probably also make it clear that core keywords in all subschemas would have to be evaluated before other keywords in the roots [or not, but at least *some* specific defined behavior is needed] -- for example, consider the modified document 1 but with "someProperty" placed *after* its sibling "newPropertyX" keys [a case I missed when building that example].)

**Reply by karenetheridge** [reply]
The schema whose `$id` is the resolved result of the `$ref` must be "known" to the evaluator at the time of the `$ref` evaluation, or the evaluation will fail (not with an "invalid" result, but a "neither valid nor invalid, but rather an error" type of result).  Whether the implementation chooses to load that resource at runtime (for example by fetching a file from disk or a document from the internet) is at the implementation's discretion -- it is not required to support just-in-time document loading:

https://json-schema.org/draft/2020-12/json-schema-core.html#section-8.2.3

> he resolved URI produced by these keywords [`$ref`, `$dynamicRef`] is not necessarily a network locator, only an identifier. A schema need not be downloadable from the address if it is a network-addressable URL, and implementations SHOULD NOT assume they should perform a network operation when they encounter a network-addressable URI.

**Reply by JC3** [reply]
Sure; but if the application chooses to load the resource at runtime, then any additional ids defined _in_ that resource necessarily can't be known ahead of time. So it seems like while the spec leaves that behavior up to the application, the application's choice (which theoretically shouldn't affect the outcome, otherwise the spec shouldn't leave it application-defined) becomes important: If it fetches at runtime then it must necessarily resolve subschemas depth-first or do some other sort of multi-pass evaluation, and if it fetches ahead of time then it won't be able to handle ids defined in referenced documents unless it actually evaluates all those references while doing the ahead-of-time fetch (so it can get all dependencies), which creates a bit of a paradox. Especially if the document it is referencing is outside of the application's control e.g. defined by a third-party.

**Comment by handrews** [comment]
@JC3

> Should any of those other reference work?

Generally speaking, regardless of how you loaded a schema resource, it should be available under its `$id`.  Whether or not it is also available by its retrieval URL is definitely not covered by JSON Schema, nor do I think it particularly needs to be.  Referencing by `$id` is the correct approach and is much more predictable than retrieval URIs.

I would strongly advise against mixing addressing modes like this.  It's the mixing that introduces a dependency on the keyword order.  If you use consistent addressing, it doesn't matter which is resolved first.  Even if there was a defined ordering, it does not seem like a good idea to rely on it to establish an alternate addressing approach.  You could easily break things by a fairly benign-seeming schema change.

If you want to load schemas from disk, do that up front.  If you're writing an implementation, one strategy I've seen is registering handlers for URI resolution, which would (for example) allow you to intercept a URI and convert it to a local filesystem path during testing, but do a network retrieval in production.

I would say it's a "best practice" to separate the identification and location of schemas.  Identification needs to remain stable, but location might change depending on the environment.  You don't want to have to edit your schemas for that.

**Reply by Relequestual** [reply]
> Generally speaking, regardless of how you loaded a schema resource, it should be available under its $id. 

100% yes.

> Whether or not it is also available by its retrieval URL is definitely not covered by JSON Schema, nor do I think it particularly needs to be.

I'm not sure that's quite correct. I think the spec DOES actually specify this [here](https://json-schema.org/draft/2020-12/json-schema-core.html#name-initial-base-uri) by referencing https://www.rfc-editor.org/rfc/rfc3986.html#section-5.1

Base URI embedded in content first. (the ASCI diagram, from inside to outside.)

In terms of best practice @JC3... [further reading](https://json-schema.org/draft/2020-12/json-schema-core.html#name-loading-a-referenced-schema)

>The use of URIs to identify remote schemas does not necessarily mean anything is downloaded, but instead JSON Schema implementations SHOULD understand ahead of time which schemas they will be using, and the URIs that identify them.

This means, implementations should provide a method for you to pre-load the schemas you'll be using, and then make an index of their identifier URI for referencing later. Of course, dynamically populating that index is possible, but as @karenetheridge points out, that might not be reliable and may even have race doncitions.

**Reply by handrews** [reply]
@Relequestual that's about determining the base URI.  The question here is if you use a `file://` URL to locate the schema resource, which has a different URI (an `xri://` one, in this example) in the `$id`, when you save the schema to allow resolving future references, it's clear that you MUST save it under the `$id`.  If there ***isn't*** a `$id`, then you'd save it under the retrieval URL.  But if there ***is*** a `$id`, do you also remember the retrieval URL and save it under that, too?

Section 9.2.1 suggests that this is not required.  The `$id` is the canonical URI, and keeping track of any others is optional.

> An implementation MAY choose not to support addressing schema resource contents by URIs using a base other than the resource's canonical URI, plus a JSON Pointer fragment relative to that base. Therefore, schema authors SHOULD NOT rely on such URIs, as using them may reduce interoperability. This is to avoid requiring implementations to keep track of a whole stack of possible base URIs and JSON Pointer fragments for each, given that all but one will be fragile if the schema resources are reorganized. Some have argued that this is easy so there is no point in forbidding it, while others have argued that it complicates schema identification and should be forbidden. Feedback on this topic is encouraged. After some discussion, we feel that we need to remove the use of "canonical" in favour of talking about JSON Pointers which reference across schema resource boundaries as undefined or even forbidden behavior (https://github.com/json-schema-org/json-schema-spec/issues/937, https://github.com/json-schema-org/json-schema-spec/issues/1183)

**Reply by JC3** [reply]
Thanks! So, I think my example might not have been the best: I only did it that way because it was the quickest way to contrive a situation where a URI was only resolvable because it was indirectly encountered in a referenced schema, which is what that was really about. 

Consider the following situation instead (I'm omitting stuff for brevity, but these are 2020-12). There might be a way to simplify this example, not sure, but maybe it's more realistic. First, three documents:

_Document A: Retrieval URL: https://important.org/complicated-device.schema.json:_

```
{
    "$id": "https://important.org/complicated-device.schema.json",
    "description": "some complicated schema, referencing a large collection of other documents",
    "properties": {
        "drivetrain": { "$ref": "components/encabulator.schema.json" },
        ... etc.
    }
}
```

_Document B: Retrieval URL: https://important.org/components/encabulator.schema.json:_

```
{
    "$id": "https://important.org/components/encabulator.schema.json",
    "$defs": {
        "ambiguator": {
            "$comment": "schema moved into encabulator, see heated usenet discussion for rationale",
            "$id": "https://important.org/components/transdimensional/ambiguator.schema.json",
            "properties": { ... }
        }
    },
    "description": "basic turbo encabulator",
    "properties": { ... }
}
```

_Document C: Retrieval URI irrelevant:_

```
{
    "$id": "xri://+random.application/schemas/data",
    "properties": {
        "device": { "$ref": "https://important.org/complicated-device.schema.json" },
        "spareTDA": {
              "$comment": "user may specify a spare transdimensional ambiguator, if necessary",
              "$ref": "https://important.org/components/transdimensional/ambiguator.schema.json"
        }
    }
}
```

Assumptions:
- Important.org is some organization with a public specification and a known URL; its documentation states that the URL is https://important.org/complicated-device.schema.json. A common scenario.
- Important.org's spec is split into multiple resources; but this is more important.org's business than the user's -- it's just how they've chosen to organize. Again, a common scenario: Provide a main resource at a known location for users to reference, which itself may reference other resources.
- https://important.org/components/transdimensional/ambiguator.schema.json is _not_ an accessible URL; there's no document there, accessing it would yield a 404. It's just a location-independent ID in _encabulator.schema.json_. Important.org put it there, determining it to be good practice after reading the JSON Schema spec's statement that location-independent IDs allow "[a subschema to be relocated without requiring JSON Pointer references to be updated](https://json-schema.org/draft/2020-12/json-schema-core.html#name-defining-location-independe)".
- Document C is some application's own schema. The application makes use of important.org's specifications.
- The application strategy for resolving missing references is to treat the URI as a URL and attempt to fetch it.
- The only schemas the application "understands ahead of time" to be available are Document C. Also, of course, anything accessible via https is fetchable at evaluation time.

So the application loads document C. It sees the reference to document A, it does its thing, resolves all references, everything works out.

Then, it sees document C's references to https://important.org/components/transdimensional/ambiguator.schema.json. 

So the question in this example, then, is: evaluation order and poor design / implementation decisions aside, should the application be able to resolve that reference without a network operation? Or should the result be a 404 presumably followed by an application error?

---

Also, something to note about both this example and my original one: Bad decisions notwithstanding, afaict nothing is in violation of the JSON Schema spec.

Is that a clearer example?

**Reply by JC3** [reply]
Also, I didn't see the replies before posting that; so just to reiterate:

>  The question here is if you use a file:// URL to locate the schema resource, which has a different URI (an xri:// one, in this example) in the $id, ...

The question I was _trying_ to ask was actually: When a `$ref` is resolved, does the referenced document become part of the evaluator's general scope (as if it were pre-loaded), or is the referenced subschema merely dropped in and then the referenced document forgotten, or somewhere in between e.g. the referenced document behaves as if it were pre-loaded but only when evaluating the subschema that the `$ref` was in, etc.

**Reply by handrews** [reply]
@JC3 yes, that's more clear, although you did not explicitly specify if _encabulator.schema.json_ is accessible by treating its `$id` as a URL, but since you gave it that URL as its retrieval URL I assume that it is indeed accessible.

Assuming that is correct, and that your implementation correctly scans `$defs` for schema objects that might use `$anchor` and/or `$id`, then all of this should work fine.  And the test suite's tests for `$anchor` and `$id` should be enough to ensure that implementations correctly scan `$defs`.

**Reply by handrews** [reply]
> When a $ref is resolved, does the referenced document become part of the evaluator's general scope (as if it were pre-loaded)

We should probably be more explicit about this in the spec, but yes, that's my expectation.

**Reply by JC3** [reply]
> > When a $ref is resolved, does the referenced document become part of the evaluator's general scope (as if it were pre-loaded)
> 
> We should probably be more explicit about this in the spec, but yes, that's my expectation.

Thanks. Then, back to evaluation order (in a sense); if document C were modified to this (the order of the two properties has been swapped):

```
{
    "$id": "xri://+random.application/schemas/data",
    "properties": {
        "spareTDA": {
              "$comment": "user may specify a spare transdimensional ambiguator, if necessary",
              "$ref": "https://important.org/components/transdimensional/ambiguator.schema.json"
        },
        "device": { "$ref": "https://important.org/complicated-device.schema.json" }
    }
}
```

Should that also be expected to work?

Note that RFC7159 defines JSON objects as "[an _unordered_ collection](https://datatracker.ietf.org/doc/html/rfc7159#section-1)" (that is, the modified document C above is technically equivalent to the original document C). It also notes that "[JSON parsing libraries have been observed to differ as to whether or not they make the ordering of object members visible to calling software](https://datatracker.ietf.org/doc/html/rfc7159#section-4)" i.e. it may not actually be possible in practice for a given JSON Schema evaluator to be aware of the ordering of keys in the documents at all.

Also note that there is no distinguishing difference between */properties/device/$ref* and */properties/spareTDA/$ref* that would allow the JSON Schema spec to concretely define the order in which they are evaluated, other than their names and the referenced URIs.

---

PS:

> you did not explicitly specify if encabulator.schema.json is accessible by treating its $id as a URL, but since you gave it that URL as its retrieval URL I assume that it is indeed accessible.

Your assumption is correct; sorry about not explicitly mentioning that!

**Reply by handrews** [reply]
@JC3 if you are relying on performing a network operation whenever you encounter `$ref`, then no, that is not guaranteed to work.

Basically, it's ill-advised to rely on network operations during processing.  But it's not just that:  in the previous example, there's nothing that forces you to process `$defs` and notice the embedded resource prior to needing it.

Section 9 on loading schemas is not as clear as I'd like it to be, in part because there's not consensus on how schemas ought to be loaded.  In my view, loading, which includes scanning the lexical scopes for schemas and noting their URIs, is a separate process that happens before validation.  While it's possible to try to find schemas on-demand, it's a lot more complex to do so, and the spec does not require handling all of the possible cases that you might come across.  Your latest example pretty much explains why:  it's not necessarily possible to figure out how to locate all of the schemas.

Typically what I see is people loading all of the schemas up front, and then using them.

**Reply by JC3** [reply]
> But it's not just that: in the previous example, there's nothing that forces you to process $defs and notice the embedded resource prior to needing it.

I see; that was a big component in my question; whether or not other embedded resources in the referenced document were put in play when a reference to some subschema in that document was resolved.

I'm a little confused now though, so just to confirm, this is no longer the answer:

> > When a $ref is resolved, does the referenced document become part of the evaluator's general scope (as if it were pre-loaded)
> 
> We should probably be more explicit about this in the spec, but yes, that's my expectation.

And instead it's:

> @JC3 if you are relying on performing a network operation whenever you encounter $ref, then no, that is not guaranteed to work.

Correct?

Note that the assumption here is that all retrieval operations that can succeed, will succeed... that is, the topic is the JSON Schema spec rather than, say, the reliability of networks (or the user's hard drive crashing for file URLs, or the user's mail server being down for cid URLs, or cosmic rays corrupting RAM with random bit flips at just the wrong time, or the possibility of a meteor destroying half the planet just when a `$ref` is about to be resolved, etc...).

---

> Typically what I see is people loading all of the schemas up front, and then using them.

In my last example, let's go back in time a little to just before important.org's infamous "heated usenet discussion", when /components/transdimensional/ambiguator.schema.json was an accessible document, via that URL.

Let's assume that the user is using that example application to validate an instance against document C. Can you give an example of:

- How important.org communicated the list of documents to preload to the application developer or user (your choice), and
- How that list is input to the application?

Now, with that in mind, go forward in time again. Important.org has now made a change to their file structure; ambiguator.schema.json is no longer network-accessible, and has been moved into encabulator.schema.json and given a location-independent `$id`. Can you give an example of:

- How the application developer (or user) is made aware that important.org changed their internal file structure, and
- How important.org communicates the new list to the application developer or user, and
- How the application developer or user proceeds to update the input to the application?

**Reply by handrews** [reply]
@JC3 sorry for dropping the ball on this, some Real Life Stuff™ came up and I'm still working through the subsequent backlog.

> Note that the assumption here is that all retrieval operations that can succeed, will succeed... that is, the topic is the JSON Schema spec rather than, say, the reliability of networks (or the user's hard drive crashing for file URLs, or the user's mail server being down for cid URLs, or cosmic rays corrupting RAM with random bit flips at just the wrong time, or the possibility of a meteor destroying half the planet just when a $ref is about to be resolved, etc...).

The JSON Schema specification explicitly avoids requiring network operations during processing.  The expectation is that schemas are pre-fetched and pre-loaded, and implementations hold on to them in some sort of URI-indexed storage.  `$ref` is resolved against that.  Some implementations offer network or other fetch capabilities, but that's additional functionality beyond what the spec requires.  

Part 2 of the presentation I posted for discussion at #220 talks more about scanning for URIs within a schema (with diagrams), and how those should be available through the URI-indexed storage once the larger schema document is loaded.

All of those other things are application-specific.  The exact interface to pre-load a schema is JSON Schema implementation-specific, but should be clearly documented for any given implementation.  But there's no way JSON Schema can define a process for all of that other stuff that would be universally well-suited to all applications.
